Add support for `spring.config.import` with S3 backend

kunalvarpe opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [x] Refactoring


## :scroll: Description
<!--- Describe your changes in detail -->
Added supported classes for config import which is done is AWS parameter source or secret-manager modules

Fixes #161

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->


## :green_heart: How did you test it?


## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [ ] All tests passing
- [ ] No breaking changes


## :crystal_ball: Next steps

# Note: The PR is still in progress state, need more discussions to procceed further.

kunalvarpe commented:
> This is only for using S3 Bucket as storage for configuration correct?

Yes, exactly the earlier scope of the ticket was to support the s3 backend. But in later discussion on issue we thought of it should be streamline with `spring.config.import`.  So below things are yet to be explore.

- S3 backend with support for respecting profiles.
- `spring.config.import` with file systems

Here I need some help and understanding and finalize the scope for developemet as part of this ticket.

MatejNedic commented:
Hey @kunalvarpe thanks on great work, I introduced some refactor since we changed underlying classes.
This will be target for new Spring Cloud AWS release!

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLoader.java:
```diff
@@ -0,0 +1,72 @@
+public class S3ConfigDataLoader implements ConfigDataLoader<S3ConfigDataResource> {
+
+	public S3ConfigDataLoader(DeferredLogFactory logFactory) {
+		BootstrapLoggingHelper.reconfigureLoggers(logFactory, "io.awspring.cloud.S3PropertySource",
```
`io.awspring.cloud.S3PropertySource` has a wrong package

kunalvarpe replied:
This has been addressed

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLocationResolver.java:
```diff
@@ -0,0 +1,131 @@
+	static class S3ClientFactory {
+
+		static S3Client s3Client(S3ClientBuilder builder) {
+			if (ClassUtils.isPresent("io.awspring.cloud.s3.crossregion.CrossRegionS3Client", null)) {
```
CrossRegionS3Client is deprecated, lets not use it anymore.

kunalvarpe replied:
This has been addressed

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLocationResolver.java:
```diff
@@ -0,0 +1,131 @@
+				loadCredentialsProperties(resolverContext.getBinder()));
+		registerBean(resolverContext, RegionProperties.class, loadRegionProperties(resolverContext.getBinder()));
+
+		registerAndPromoteBean(resolverContext, S3Client.class, BootstrapRegistry.InstanceSupplier.of(
```
If it is a supplier, perhaps it is better to pass a lambda there? 

```
BootstrapRegistry.InstanceSupplier.of(() -> ...);
```

kunalvarpe replied:
This has been addressed

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3PropertySources.java:
```diff
@@ -0,0 +1,71 @@
+				throw new AwsS3PropertySourceNotFoundException(e);
+			}
+			else {
+				LOG.warn("Unable to load AWS S3 object from " + context + ". " + e.getMessage());
```
shouldn't we always log warning, like we do in Parameter Store and Secrets Manager?

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/**
+ * Auto-configuration for S3 integrations.
```
```
{@link org.springframework.boot.context.config.ConfigDataLoader} implementation for AWS S3.
```

kunalvarpe replied:
This has been addressed

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/properties/S3Properties.java:
```diff
@@ -93,6 +94,25 @@ public class S3Properties extends AwsClientProperties {
+	/**
+	 * Enables S3 integration for spring.config.import.
+	 */
+	private boolean enableImport = true;
+
+	/**
+	 * Properties related to configuration reload.
+	 */
+	@NestedConfigurationProperty
+	private ReloadProperties reload = new ReloadProperties();
```
i think these two properties should not be at the same level as the rest as they apply specifically to config import feature. Perhaps we can move them to inner class `Config`? What do you think?

kunalvarpe replied:
I think this make sense. Lets confirm so that I can address this comment.

maciejwalkowiak commented on spring-cloud-aws-s3-parent/spring-cloud-aws-s3-config/pom.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
do we need this in separate module? Perhaps it can be added to S3 module?

kunalvarpe replied:
@maciejwalkowiak @MatejNedic 

I am not sure but for S3 only we have made separate module for config and there is only one class there that is `S2PropertySource`.  Not sure intention behind adding new module. If not need we can move both pom and above class to S3 module under S3-Parent.

kunalvarpe commented:
> Hey @kunalvarpe thanks on great work, I introduced some refactor since we changed underlying classes. This will be target for new Spring Cloud AWS release!

@MatejNedic Thank you so much for your support.

kunalvarpe commented:
@MatejNedic @maciejwalkowiak 

I have addressed the comments on which I have commented. Please review.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/parameterstore/package-info.java:
```diff
@@ -15,7 +15,7 @@
  */
 
 /**
- * {@link org.springframework.boot.context.config.ConfigDataLoader} implementation for AWS Parameter Store.
```
there's something off here

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLoader.java:
```diff
@@ -0,0 +1,72 @@
+	public ConfigData load(ConfigDataLoaderContext context, S3ConfigDataResource resource) {
+		try {
+			// resource is disabled if s3 integration is disabled via
+			// spring.cloud.aws.s3.enable_import=false
```
this does not seem right. shouldn't be `spring.cloud.aws.s3.config.enabled=false`?

MatejNedic replied:
It should didn't check it

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/properties/S3Properties.java:
```diff
@@ -175,4 +197,12 @@ public S3Configuration toS3Configuration() {
 		return config.build();
 	}
+
+	public boolean isEnableImport() {
```
there is already `config.enabled` - isn't this redundant?

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLoaderIntegrationTests.java:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	@Test
+	void resolvesPropertyFromS3CompplexPath() {
```
`compplex` -> `complex`

maciejwalkowiak commented on spring-cloud-aws-starters/spring-cloud-aws-starter-s3-config/pom.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
do we need a dedicated starter for it? or should just s3 started be used?

MatejNedic replied:
I think this is bug in IntelliJ I don't see this file

maciejwalkowiak commented on docs/src/main/asciidoc/s3.adoc:
```diff
@@ -254,6 +254,246 @@ The Spring Boot Starter for S3 provides the following configuration options:
+|===
+| Name | Description | Required | Default value
+| `spring.cloud.aws.s3.config.enabled` | Enables the S3 config import integration. | No | `true`
+| `spring.cloud.aws.s3.config.endpoint` | Configures endpoint used by `S3Client`. | No | `null`
```
these properties do not exist as far as i can tell. should config import have different endpoint/region etc properties than regular S3 integration?

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/AwsS3ClientCustomizer.java:
```diff
@@ -0,0 +1,26 @@
+ * @author Kunal Varpe
+ * @since 3.2.0
+ */
+public interface AwsS3ClientCustomizer extends AwsClientCustomizer<S3ClientBuilder> {
```
This will have to be changed if #1229 gets merged first.

MatejNedic commented:
@maciejwalkowiak I think we are good now can you please review?

maciejwalkowiak commented on docs/src/main/asciidoc/s3.adoc:
```diff
@@ -378,6 +378,244 @@ There can be multiple `S3ClientCustomizer` beans present in single application c
+
+	@Override
+	public void initialize(BootstrapRegistry registry) {
+		registry.register(AwsS3ClientCustomizer.class,
```
`AwsS3ClientCustomizer` is deprecated. Instead we should show in examples usages of `S3ClientCustomizer`

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ManagerClientCustomizer.java:
```diff
@@ -0,0 +1,15 @@
+ * @since 3.4.0
+ */
+@FunctionalInterface
+public interface S3ManagerClientCustomizer  extends AwsClientCustomizer<S3ClientBuilder> {
```
Do we need a separate customizer here? Can't users use `S3ClientCustomizer`?

MatejNedic commented:
After long time feels nice to merge this. Thanks @kunalvarpe on PR!
